### PR TITLE
Apply a threadpool to fix the OOM issue of Logstash

### PR DESCRIPTION
**Pull request by ph**
Add a simple threadpool to limit the number of connection to the current lumberjack server.

fixes: https://github.com/logstash-plugins/logstash-input-lumberjack/issues/10

### CODE REVIEW

**Comment by ph** [comment]
@colinsurprenant @jsvd  review please?

**Comment by suyograo** [comment]
@jordansissel review as well :)

File: lib/logstash/inputs/lumberjack.rb [hunk 1]
```diff
@@ -29,19 +29,37 @@ class LogStash::Inputs::Lumberjack < LogStash::Inputs::Base
   # SSL key passphrase to use.
   config :ssl_key_passphrase, :validate => :password
 
+  # The lumberjack input using a fixed thread pool to do the actual work and
+  # will accept a number of client in a queue, before starting to refuse new
+  # connection. This solve an issue when logstash-forwarder clients are 
+  # trying to connect to logstash which have a blocked pipeline and will 
+  # make logstash crash with an out of memory exception.
+  config :max_clients, :validate => :number, :default => 1000
```
Comment (jordansissel) [inline lib/logstash/inputs/lumberjack.rb]: If this is just the size of the accept queue, then I think the name `max_clients` is a misnomer. 

I also wouldn't necessarily note the bug this setting fixes in the documentation.

Comment (ph) [inline lib/logstash/inputs/lumberjack.rb]: Okay, I can remove the details.

`@max_accept_queue` ?;)

Comment (jordansissel) [inline lib/logstash/inputs/lumberjack.rb]: When I finish reviewing+testing I will hopefully have a suggestion :P

Comment (ph) [inline lib/logstash/inputs/lumberjack.rb]: Do you have an idea, to define the default value for it?
Do we have some data we could use to give it a sane default?

**Comment by allenmchan** [comment]
I am trying to understand how this impacts my environment. 
Based on https://github.com/logstash-plugins/logstash-input-lumberjack/pull/5, i believe each lumberjack connection would create it's own thread. 
i would need to set this proposed threadpool equal to the max amount of logstash-forwarders i have correct?

**Comment by ph** [comment]
@allenmchan Correct,  but dont use it yet, after discussing with @jordansissel, I'll experiment with another way. I've changed the title of this PR to WIP

File: lib/logstash/inputs/lumberjack.rb [hunk 2]
```diff
@@ -29,19 +29,33 @@ class LogStash::Inputs::Lumberjack < LogStash::Inputs::Base
   # SSL key passphrase to use.
   config :ssl_key_passphrase, :validate => :password
 
+  # Number of maximum clients that the lumberjack input will accept
```
Comment (suyograo) [inline lib/logstash/inputs/lumberjack.rb]: @ph can you add why we are immediately deprecating this setting in the comments. Will be useful for anybody wondering...

**Comment by allenmchan** [comment]
So i have a lot more than 100 clients. How do i configure max_clients to the number i want?

**Pull request closed without merging**